### ISSUE TITLE
AssitiveMML extension: visual rendering not aria-hidden in IE10+ with NVDA

### ISSUE THREAD

**Issue by pkra** [mathjax/MathJax#1235]
Just to keep this separate from #938

**Comment by pkra** [comment]
This seems to be triggered by NVDA? If NVDA is not active, the aria-hidden attribute is present. cc @jcsteh

I may just be interpreting things incorrectly. The inspector shows `="true` and the visual output is ignored (but the MathML not read correctly). Maybe this is expected behavior?

**Comment by dpvc** [comment]
Here is some more information:  this doesn't seem to be MathJax-specific, but happens in static pages as well.  For example, with this file

``` html
<!DOCTYPE html>
<html>
<head>
<title>NVDA aria-hidden test</title>
</head>
<body>

The following has aria-hidden set to true:
<div aria-hidden="true">skip me</div>
It should not have been read.

</body>
</html>
```

if I have NVDA running, then open this file in IE10 on Windows 7, the page will read properly the first time (the text of the `<div>` will be skipped), but if you move the mouse and cause it to be read again, the "skip me" *is* read.  Further, if you use the developer tools to inspect the live DOM, the `aria-hidden="true"` attribute has been removed!  Steps to reproduce:

*  Start NVDA
*  Open the page above in IE10
*  NVDA reads the page
*  Open Developer Tools (F12)
*  In the developer window, open the `html` and `body` elements so you can see the `div` element.  It will be missing the `aria-hidden` attribute.

**Comment by dpvc** [comment]
Further information:  Even when the developer tools don't show the `aria-hidden` attribute, inspecting the DOM through the console (say via `document.body.childNodes[1].getAttribute("aria-hidden")` or `document.body.childNodes[1].outerHTML` or `document.body.innerHTML`) shows it is set.  But the fact that the developer tools don't show that does suggest that something funny is going on.  That, plus the fact that NVDA reads the node anyway.

**Comment by dpvc** [comment]
Yet more information:  A similar things happens with MathML in the page in IE10.  With MathPlayer4 installed, the page

``` html
<!DOCTYPE html>
<html>
<head>
<title>NVDA MathML test</title>
</head>
<body>

A fraction: 
<math>
  <mfrac>
    <mi>x</mi>
    <mn>2</mn>
  </mfrac>
</math>

</body>
</html>
```

will be read properly the first time it is read, but moving the mouse to get it to read a second time and the math will be read "x 2" rather than "the fraction with numerator x and denominator 2".

**Comment by pkra** [comment]
Have you tried removing MathPlayer? That seems to get me to NVDA just reading the string from the MathML (and ignoring the visual output), cf. https://github.com/mathjax/MathJax-docs/issues/111

**Comment by dpvc** [comment]
I started without MathPlayer, and yes, it reads the MathML by ignoring the tags.  To get it to read it as math, you need to have MathPlayer installed.  The point here is that even *with* MathPlayer, the second time is read as text rather than math.

**Comment by pkra** [comment]
Weird. 

FYI, I've filed http://community.nvda-project.org/ticket/5283#ticket

**Comment by pkra** [comment]
> I started without MathPlayer, and yes, it reads the MathML by ignoring the tags. To get it to read it as math, you need to have MathPlayer installed.

Ah, ok. I didn't understand it that way. As mentioned on the other thread, I had trouble with MathPlayer installed but uninstalling it fixed those. Maybe uninstalling-but-not-rebooting helped eliminated the MathPlayer problems while leaving enough of the library to run NVDA? Super weird.

**Comment by jcsteh** [comment]
Aria-hidden and math content may not be handled correctly in IE when reading with the mouse. Although idealy mouse reading would work, it's not a common use case for a screen reader user, so it's lower priority for us (and much harder to get right). Please try with the keyboard. See the NVDA User Guide section regarding Browse Mode.

**Comment by pkra** [comment]
Thanks for the feedback! Could you look at my comments on the other thread
and the docs wiki issue. Keyboard usage also does not read correctly.

**Comment by pkra** [comment]
@dpvc should we dropped this from the 2.6 milestone? Seems like we can't work on a fix without assistance from nvaccess.

**Comment by dpvc** [comment]
> should we drop this from the 2.6 milestone?

Yes, I think we can't do anything about it now.

**Comment by pkra** [comment]
I was just looking this up.

* the first example (just aria-hidden) seems fixed in current NVDA.
* the problem with MathML being read both as "string" and correctly persists.

I'll update my ticket at nvaccess which now resides at https://github.com/nvaccess/nvda/issues/5283.

**Comment by pkra** [comment]
So that's embarrassing. I cannot reproduce the plain MathML/MathPlayer problem (no MathJax involved). @dpvc if you can still reproduce this (and even after rebooting the VM), could you let me know?

I still have the original issue of MathJax's AssistiveMML extension leading to double reading, e.g., here http://s.codepen.io/pkra/debug/NGojag, "x 2 x over 2".

**Comment by pkra** [comment]
@dpvc: at https://github.com/nvaccess/nvda/issues/5283#issuecomment-158271803, @jcsteh (thank you!) dug into this for us and explains how this happens via bad IE behavior (surprise!) triggering NVDA heuristics. 

A fix from our end could be to add `role=presentation` on both the output span (where we have `role=math`) and the `<math>` elements parent span. We should test how other AT would be affected by such a change and I'll ask around. 

I don't think any AT I tested announced the math role in IE anyway, so even if `role=presentation` turns out not to be a good global solution, we can perhaps still use it for IE.

**Comment by pkra** [comment]
Just a reverse ping https://github.com/nvaccess/nvda/issues/5283#issuecomment-160193109.  I've made a hacked copy of the extension in the pen at http://codepen.io/pkra/pen/NGojag?editors=101. 

* add role=presentation to MathJax's top-level span
* add role=presentation to parent span of the MathML.

NVDA works fine, JAWS, too, but drops the "clickable math" since the menu is attached to the top-level span (correct?).  I don't know if we should mind that, actually, given how many users complain about JAWS voicing "clickable math" after every equation.

**Comment by pkra** [comment]
The results were only for IE10 on Win8.1. Will test more.

**Comment by pkra** [comment]
I've tested all platforms we've tested before. The change to role=presentation works across the board. 

The only "issues" are

* no more "clickable" etc. indicating the menu. But since users complained, I think this is actually positive.
* VoiceOver iOS does not have an off-set highlight around the hidden MathML. That means you can't double click on it to start Safari's MathML explorer. I couldn't find a reference on how to start that explorer with a gesture. So that would be the only real downside I can see. For me, the gain outweighs it.

**Comment by jcsteh** [comment]
For what it's worth, I still consider nvaccess/nvda#5283 to be a problem that we need to fix in NVDA (or rather, yet another IE stupidity that we need to hack around in NVDA). Unfortunately, I don't have an ETA for this; it's a nontrivial fix.

> no more "clickable" etc. indicating the menu. But since users complained, I think this is actually positive.

IMO, clickable isn't a good indication of a menu anyway. If anything, you want to use aria-haspopup (but that doesn't give an indication of how to activate it).

> VoiceOver iOS does not have an off-set highlight around the hidden MathML. That means you can't double click on it to start Safari's MathML explorer.

I assume you can still flick to get to the math, though? Otherwise, how do you read it? Is it possible to make the hidden MathML have a greater width and height as far as browsers are concerned without actually making it visible? That might fix the Safari issue and maybe the menu targetability as well.

**Comment by jcsteh** [comment]
Ideally, the hidden math would have screen coordinates which fall within the wrapper. This way, if ATs try to click the hidden math, they'll hit the wrapper, which is exactly what you want. It should also cause the wrapper to be highlighted if the AT highlights the coordinates of the hidden math, which is also what you want. I'm not sure if this is possible, though.

**Comment by pkra** [comment]
> For what it's worth, I still consider nvaccess/nvda#5283 to be a problem that we need to fix in NVDA (or rather, yet another IE stupidity that we need to hack around in NVDA). Unfortunately, I don't have an ETA for this; it's a nontrivial fix.

Thanks, that's good to know. Still, it seems to me that switching to role=presentation would be a reasonable tradeoff until this is fixed in NVDA. Especially, since I checked our old solution and I don't see the outline on the hidden MathML in iOS anymore. I'm guessing an iOS update may have changed that behavior.

> I assume you can still flick to get to the math, though?

Yes, I was flicking.

> Is it possible to make the hidden MathML have a greater width and height as far as browsers are concerned without actually making it visible? 

That's an interesting idea but I think that's out of scope for 2.6. It would also come with some performance issues.

> Ideally, the hidden math would have screen coordinates which fall within the wrapper. 

Hm. I thought that's the case. @dpvc?

What worries me slightly is your comment at https://github.com/nvaccess/nvda/issues/5283#issuecomment-161160096. I hope it's ok to move my response to this thread where it fits better, I think.

> I was trying to route the mouse to the node and right click. I tried both the outer wrapper and the hidden math. Testing again, it seems to work in IE, but not Firefox.

I'm not sure what you mean by routing here (but see below). Also, I'm not sure if "testing again" refers to the old implementation (aria-hidden, aria read-only) or new implementation (role presentation => aria hidden, role presentation).

I did a quick check of  the following: tabbing to an equation, moving mouse to current navigator object (NVDA+numpadDivide / NVDA+shift+m resp.), and triggering right-click (numpadDivide / NVDA+[ resp.). I wasn't able to reliably get to the menu in the old/current AssistiveMML extension, because the cursor isn't always on target. E.g., on mathjax.org where the old/current implementation is active, the first equation does not work but in the samples in the first modal the cursor is on target. 

However, I also couldn't get the menu in the new implementation at http://s.codepen.io/pkra/debug/NGojag, again the cursor would be off.

> I didn't know about [using space to get the menu in v2.6]. It does work nicely, but the problem is that the wrapper has to get focus for that to work and screen reader users have to switch to focus mode or pass the key through (since space/enter clicks in browse mode/equivalent).
> Unfortunately, the wrapper never gets focus now and that's due to role="presentation". Without role="presentation", NVDA does focus the wrapper when you move to the math. This probably isn't that useful anyway due to the space/enter thing I mentioned, but it is the sort of regression I was worried about.

I didn't realize NVDA did anything different in the old/current AssistiveMML implementation. I can tab to the expression and trigger the menu (in focus mode) in both. I did not notice NVDA doing anything special when the expression is in focus but I thought that hasnot changed. I'm guessing my inexperience as an AT user together with my sight makes me miss something :-( . Could you possibly explain the difference in more detail?

**Comment by jcsteh** [comment]
>     Ideally, the hidden math would have screen coordinates which fall within the wrapper.
> 
> Hm. I thought that's the case. @dpvc?

They partially fall within the wrapper, but not entirely. The top of the hidden math falls within the wrapper, but the centre of the hidden math doesn't.

Using your CodePen test, here's what I get n my system for the screen locations in Firefox:

* Wrapper (mjx-chtml): left 102, top 135, right 118, bottom 162
* Hidden "math" element: left 102, top 161, right 113, bottom 182
* Centre of hidden "math" element: 107, 171 = 10 pixels below the bottom of the wrapper

>     I was trying to route the mouse to the node and right click. I tried both the outer wrapper and the hidden math. Testing again, it seems to work in IE, but not Firefox.
> 
> I'm not sure what you mean by routing here

You got my meaning; I meant routing the mouse as you describe.

> Also, I'm not sure if "testing again" refers to the old implementation (aria-hidden, aria read-only) or new implementation (role presentation => aria hidden, role presentation).

Both. I must have tested incorrectly the first time.

> I did a quick check of the following: tabbing to an equation, moving mouse to current navigator object (NVDA+numpadDivide / NVDA+shift+m resp.), and triggering right-click (numpadDivide / NVDA+[ resp.). I wasn't able to reliably get to the menu in the old/current AssistiveMML extension, because the cursor isn't always on target. E.g., on mathjax.org where the old/current implementation is active, the first equation does not work but in the samples in the first modal the cursor is on target.

Right. Note that this will try to route to the hidden math. This is what I was seeing and is probably due to the coordinates problem above.

I just discovered that my test with the wrapper wasn't quite correct. If I get to the wrapper and ensure I'm routing to the centre of that, I can get it to activate. However, this requires some pretty advanced knowledge of NVDA and most users won't be able to pull it off. (Apparently, even i screwed it up when I tried yesterday.)

> I didn't realize NVDA did anything different in the old/current AssistiveMML implementation. I can tab to the expression and trigger the menu (in focus mode) in both.

Correct, but previously, NVDA would have focused the wrapper as soon as the browse mode cursor moved onto the math. If you then switched to focus mode or passed the next key through and pressed space, you could activate the menu. As you say, you can still tab to it and do this, but this isn't obvious.

So, in short, it'd be fantastic if we could get the centre coordinate of the math to fall within the wrapper so routing the mouse and right clicking would at least work.

What does left click currently do? If we could get that bringing up the menu, an AT user could simply hit enter. I realise this might conflict with something else, though.

**Comment by pkra** [comment]
Thanks for such a detailed follow-up, @jcsteh! We will look into this in more detail.

**Comment by dpvc** [comment]
I have made some proposed changes in the `issue1235` branch of my fork of MathJax.  It now places the assistive-MathML container in the upper left corner of the MathJax typeset math wrapper.  Ideally, the MathML is roughly the same size as the MathJax math, so the two bounding boxes should be similar, and so the assistive MathML will be inside the MathJax wrapper (or mostly so), at least with browsers that handle some MathML.  For those that just show MathML as tag soup, the two bounding boxes may not coincide very well (and for really long equations, the center of the MathML might be outside the MathJax wrapper).

For the HTML-CSS output, the "upper left" is the upper left of the containing span, and that means that its height is the standard line height, not the height of the Math itself, as it is with the CHTML and SVG output.  So that might mean the MathML extends off the bottom of the container.

In any case, give this a try and see if that is an improvement.

**Comment by pkra** [comment]
Great! I've updated http://codepen.io/pkra/pen/NGojag. It looks like this helps (at least I now reliably hit the expression when doing mouse interactions through NVDA).

@jcsteh if you have a chance to also take a look, that would be great.

**Comment by jcsteh** [comment]
Seems to work quite nicely for me now in Firefox. Thanks! I can route 
the mouse to the math and right click to bring up the context menu, and 
since the context menu has lovely ARIA on it, it behaves exactly as it 
should.

Unfortunately, the routing still seems to be off with IE 11. I didn't 
have time to actually work out what's going on, though.

**Comment by pkra** [comment]
Right, I think @dpvc had another idea to try.

**Comment by dpvc** [comment]
I already tried it and it doesn't help.  I will have a look at IE11 to see if I can see anything else to try.

**Comment by pkra** [comment]
Ah, ok.

Also, totally forgot to than @jcsteh for the positive feedback -- very much appreciated!
